Add support for favorite/pin dags to dashboard

PNL0 opened the pull request:
The DAGs can be favorited in the DAGs list menu and are displayed in the Dashboard for easier access.
Information about the favorited DAGs is stored by user in the database and retrieved with API calls. 

Dashboard view with 3 favorited DAGs (2 of them with previous DAG runs):

![image](https://github.com/user-attachments/assets/b7b65b51-fa67-4048-8c5a-79e7762c7c68)

DAGs list menu view with some of the DAGs favorited:

![image](https://github.com/user-attachments/assets/4301f1dc-bba7-4af8-bce5-3809011cc28e)

cc @pierrejeambrun @bbovenzi 

Closes: #45207

PNL0 committed:
Add support for favorite/pin dags to dashboard, using localStorage.

Co-authored-by: Mariana Isabel Marçal Santana <mariana.marcal.santana@tecnico.ulisboa.pt>

PNL0 committed:
Add support for favorite/pin dags to dashboard, persistent between sessions.

Co-authored-by: Mariana Isabel Marçal Santana mariana.marcal.santana@tecnico.ulisboa.pt

PNL0 committed:
Implement unit tests

Co-authored-by: Mariana Isabel Marçal Santana <mariana.marcal.santana@tecnico.ulisboa.pt>

PNL0 committed:
Add support for favorite/pin dags to dashboard.

The favorited DAGs by an user, are stored persistently in the database.

Co-authored-by: Mariana Isabel Marçal Santana <mariana.marcal.santana@tecnico.ulisboa.pt>

PNL0 committed:
Added no favorites message.
Fix deleted DAGs being shown as favorites.
Review the display structure of the favorite DAGs in the dashboard.

Co-authored-by: Mariana Isabel Marçal Santana <mariana.marcal.santana@tecnico.ulisboa.pt>

PNL0 committed:
Rebase and minor fixes.

Co-authored-by: Mariana Isabel Marçal Santana <mariana.marcal.santana@tecnico.ulisboa.pt>

jscheffl commented on airflow-core/src/airflow/ui/src/pages/Dashboard/FavoriteDags/FavoriteDags.tsx:
```diff
@@ -0,0 +1,94 @@
+
+      {favorites.dags.length === 0 ? (
+        <Text color="gray.500" fontSize="sm" ml={1}>
+          No favorites yet. Click the star icon next to a DAG in the list to add it to your favorites.
```
As we are in the process of translating the UI to other languages - can you please use the i18n module and add the new texts to translated properties? See https://github.com/apache/airflow/issues/50861

PNL0 replied:
I just made a new commit to use the i18n module. 
Let me know if anything is still missing.

jscheffl reviewed (commented):
Looks good in general in my view. Besides th i18n also DB migration is missing allowing to upgrade existing setups.

jason810496 reviewed (commented):
Thanks for the PR, LGMT.

FYI: https://github.com/apache/airflow/blob/main/contributing-docs/14_metadata_database_updates.rst for fixing database migrations

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -54,23 +54,45 @@
+    return DAGCollectionResponse(
+        dags=dags,
+        total_entries=len(dags),
+    )
```
I think we are missing filtering the dag_id by the `readable_dag_ids`

mariana-marcal-santana replied:
> I think we are missing filtering the dag_id by the `readable_dag_ids`

Since this endpoint has been merged in the GetDags endpoint, we believe this has been fixed.
Thank you

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -54,23 +54,45 @@
 dags_router = AirflowRouter(tags=["DAG"], prefix="/dags")
 
 
+@dags_router.get("/favorite", dependencies=[Depends(requires_access_dag(method="GET"))])
```
not sure about `dags/favorite` maybe that should be on it's own root path `/favorites`. (also always plural). Or even better, maybe just a filter option in the `get_dags` endpoint (probably only the UI one). `favorite: true/false`. So we can beneficiate from the full features of the get_dags endpoint but also on favorites. (for instance GET all the favorite dags with a dag_id_pattern that is .... ordering_by something else)

Maybe that sould be actions on the dag. `dags/{dag_id}/favorite` `dags/{dag_id}/unfavorite`. (POST on both) So we don't even need a body. To mark a dag as favorite or unfavorite.

bbovenzi replied:
+1 on filtering the existing dags list by `favorite: bool`

mariana-marcal-santana replied:
> not sure about `dags/favorite` maybe that should be on it's own root path `/favorites`. (also always plural). Or even better, maybe just a filter option in the `get_dags` endpoint (probably only the UI one). `favorite: true/false`. So we can beneficiate from the full features of the get_dags endpoint but also on favorites. (for instance GET all the favorite dags with a dag_id_pattern that is .... ordering_by something else)
> 
> Maybe that sould be actions on the dag. `dags/{dag_id}/favorite` `dags/{dag_id}/unfavorite`. (POST on both) So we don't even need a body. To mark a dag as favorite or unfavorite.

@pierrejeambrun 

The suggested changes are in our latest commit. 
Could you give us some feedback if possible?

Thank you

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -54,23 +54,45 @@
 
+@dags_router.get("/favorite", dependencies=[Depends(requires_access_dag(method="GET"))])
+def get_favorite_dags(session: SessionDep, user: GetUserDep) -> DAGCollectionResponse:
+    """Get DAGs favorited by the user."""
```
```suggestion
    """Get DAGs favorited by the user."""
```
Reword, `favorited` seems weird

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -346,6 +368,39 @@ def patch_dags(
+    session: SessionDep,
+    user: GetUserDep,
+) -> DAGResponse:
+    """Favorite the specific DAG."""
```
```suggestion
    """Mark the DAG as favorite."""
```

mariana-marcal-santana replied:
Done!

pierrejeambrun reviewed (commented):
Nice!

A few comments

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/Dashboard/FavoriteDags/FavoriteDagCard.tsx:
```diff
@@ -0,0 +1,67 @@
+                fontSize="sm"
+                textAlign="left"
+              >
+                {dagId}
```
Let's show the dag_display_name when possible

PNL0 replied:
Done.

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/Dashboard/FavoriteDags/FavoriteDags.tsx:
```diff
@@ -0,0 +1,94 @@
+              <Text
+                _hover={{ textDecoration: "underline" }}
+                as="span"
+                color="blue.500"
```
```suggestion
                color="fg.info"
```

lets use semantic colors

PNL0 replied:
Done

bbovenzi commented on airflow-core/src/airflow/ui/src/components/DagActions/FavoriteDagButton.tsx:
```diff
@@ -0,0 +1,72 @@
+
+  return (
+    <Box>
+      <ActionButton
```
Let's add this button to the Dag/Header.tsx too

PNL0 replied:
Done.

PNL0 committed:
Use i18n module and fix DB migration.
Co-authored-by: Mariana Isabel Marçal Santana <mariana.marcal.santana@tecnico.ulisboa.pt>

PNL0 committed:
Review API endpoints

Co-authored-by: Mariana Isabel Marçal Santana <mariana.marcal.santana@tecnico.ulisboa.pt>

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -346,6 +358,55 @@ def patch_dags(
+
+    user_id = user.get_id()
+    session.execute(insert(DagFavorite).values(dag_id=dag_id, user_id=user_id))
+    session.commit()
```
No need to commit, SessionDep will do that for you when the request end.

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -346,6 +358,55 @@ def patch_dags(
+@dags_router.post(
+    "/{dag_id}/favorite",
+    responses=create_openapi_http_exception_doc([status.HTTP_404_NOT_FOUND]),
+    dependencies=[Depends(requires_access_dag(method="POST")), Depends(action_logging())],
```
Maybe we can consider that any user that can `read a dag` can create its own `favorite` of it? And therefore we only need the `GET` permission on the DAG to be able to create a favorite of it?

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/datamodels/dags.py:
```diff
@@ -121,6 +121,12 @@ class DAGPatchBody(StrictBaseModel):
 
 
+class DAGFavoriteBody(StrictBaseModel):
+    """Dag Serializer for updatable bodies."""
```
Docstring is not up to date, also this seems unused and should be deleted.

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/tests/unit/api_fastapi/core_api/routes/public/test_dags.py:
```diff
@@ -392,6 +420,74 @@ def test_patch_dags_should_response_403(self, unauthorized_test_client):
+        response = unauthenticated_test_client.post(f"/dags/{DAG1_ID}/favorite")
+        assert response.status_code == 401
+
+    def test_favorite_dag_should_response_403(self, unauthorized_test_client):
+        response = unauthorized_test_client.post(f"/dags/{DAG1_ID}/favorite")
+        assert response.status_code == 403
+
```
Can we get a test for favoriting a dag that is already a favorite ? This should raise a 409 conflict error.

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -113,9 +115,19 @@ def get_dags(
     readable_dags_filter: ReadableDagsFilterDep,
     session: SessionDep,
+    user: GetUserDep,
+    favorites: Annotated[bool | None, Query()] = None,
```
```suggestion
    is_favorite: Annotated[bool | None, Query()] = None,
```

I think we should create a reusable param out of this. Just to not clutter the `route` file, and also because it will be reusable.

Basically create a parameter, you can take a look at `common/parameters.py`. 
This will be more or less the `to_orm` method. Also I should be able to pass `False` to retrieve non favorited dag only. (if a value is defined for the filter, do the join and the filter, if it's None, don't do anything)
```python
            .join(DagFavorite, DagModel.dag_id == DagFavorite.dag_id)
            .where(DagFavorite.user_id == user_id)
```

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -346,6 +358,55 @@ def patch_dags(
+            DagFavorite.user_id == user_id,
+        )
+    )
+    session.commit()
```
same about commit.

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -346,6 +358,55 @@ def patch_dags(
+    )
+    session.commit()
+
+    return DAGResponse.model_validate(dag)
```
We return a dag response, but this does not hold any information about the `favorite/non favorite` state of the Dag, not sure if it's relevant.

PNL0 replied:
> We return a dag response, but this does not hold any information about the `favorite/non favorite` state of the Dag, not sure if it's relevant.

With the current implementation, I don't think there is the need to return specific information, but maybe I'm missing something. 
What would you suggest? I was thinking that maybe we could just return the 200 successful code, if no additional information is needed to return. Similar to what is done in the delete action, where the 204 code is returned.

pierrejeambrun replied:
Yes maybe a 204 no content could be enough.

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Dashboard/FavoriteDags/FavoriteDags.tsx:
```diff
@@ -0,0 +1,96 @@
+  const location = useLocation();
+
+  const { t: translate } = useTranslation("dashboard");
+  const { data: favorites } = useDagServiceGetDags({ favorites: true });
```
The rest of the UI is using `useDagServiceGetDagsUi` to fetch dags. The UI should probably only use this, and we should also update the `ui/get_dags` endpoint to be able to do the same filtering on favorites. (therefore a common `parameters.py` will be helpful)

PNL0 replied:
> The rest of the UI is using `useDagServiceGetDagsUi` to fetch dags. The UI should probably only use this, and we should also update the `ui/get_dags` endpoint to be able to do the same filtering on favorites. (therefore a common `parameters.py` will be helpful)

Currently working on this. 
I will submit in the next few days another commit with the necessary changes (it will probably involve a rebase, since I don't have the **useDagServiceGetDagsUi** implemented in my local version).

PNL0 replied:
Done.

pierrejeambrun reviewed (commented):
Nice, thanks for all the hard work.

A few minor suggestions, but looking good overall.

PNL0 committed:
Develop favorite filter.
Create new tests for the favorite/unfavorite endpoints.
Add favorite button to the dag header. Other small fixes.

Co-authored-by: Mariana Isabel Marçal Santana <mariana.marcal.santana@tecnico.ulisboa.pt>

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -113,9 +116,12 @@ def get_dags(
 ) -> DAGCollectionResponse:
     """Get all DAGs."""
-    query = select(DagModel)
+    is_favorite.user_id = user.get_id()
```
You don't need to do that, `QueryFavoriteFilter` can access the dependency `GetUserDep`, just declare this in the function parameter that needs it.

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -113,9 +116,12 @@ def get_dags(
     """Get all DAGs."""
-    query = select(DagModel)
+    is_favorite.user_id = user.get_id()
+    query = is_favorite.to_orm(select(DagModel))
```
Why do we perform a manual call to the `to_orm` parameter?

Passing the filter to `paginated_select` should do exactly that.

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -346,6 +352,64 @@ def patch_dags(
+    user_id = user.get_id()
+    session.execute(insert(DagFavorite).values(dag_id=dag_id, user_id=user_id))
+
+    return DAGResponse.model_validate(dag)
```
lets try no content 204.

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Dashboard/FavoriteDags/FavoriteDags.tsx:
```diff
@@ -0,0 +1,96 @@
+  const location = useLocation();
+
+  const { t: translate } = useTranslation("dashboard");
+  const { data: favorites } = useDagServiceGetDags({ isFavorite: true });
```
The UI needs to fetch dags with `useDagServiceGetDagsUi`. All other places where we need that are using the UI version. (and things are cached). (This is why you also need to propagate your backend change to the `ui` endpoint. `def get_dags` in the `ui/dags.py` file.

pierrejeambrun replied:
Also the endpoint is paginated, so it's not really a `showAll` but a show 'first page'. Maybe limit it to 10, and 'show' 10 first favorite dags.

If you want to show all you will need a scroll infinite, you can take exemple in `useDagTagsInfinite`, but I recommend to do that in a follow up.

PNL0 replied:
All done, except the follow-up.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Dashboard/FavoriteDags/FavoriteDags.tsx:
```diff
@@ -0,0 +1,96 @@
+
+import { FavoriteDagCard } from "./FavoriteDagCard";
+
+const MAX_VISIBLE = 5;
```
You don't need to do that slicing in the front-end. Endpoint is paginated, just request the first 5.

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/tests/unit/api_fastapi/core_api/routes/public/test_dags.py:
```diff
@@ -392,6 +420,90 @@ def test_patch_dags_should_response_403(self, unauthorized_test_client):
+        session.execute(
+            delete(DagFavorite).where(DagFavorite.dag_id == DAG1_ID, DagFavorite.user_id == "test")
+        )
+        session.commit()
```
To avoid having to do this and have test interfere with each other, you can take a look at the ` def _clear_db`. You'll need to add the removal of dag favorite to the `clear_db_dags`. This way you are sure all tests that `clear_db_dags` at the settup will start with a fresh clean favorite table.

PNL0 replied:
Done.

pierrejeambrun reviewed (commented):
Nice!

We're getting closer, a few suggestions

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/DagsList/DagsList.tsx:
```diff
@@ -33,6 +33,7 @@ import { useLocalStorage } from "usehooks-ts";
 
 import type { DagRunState, DAGWithLatestDagRunsResponse } from "openapi/requests/types.gen";
 import DeleteDagButton from "src/components/DagActions/DeleteDagButton";
+import { FavoriteDagButton } from "src/components/DagActions/FavoriteDagButton";
```
We also probably need a new 'filter' in case the user only want to see, list and filter his favorite dags. (A button to toggle/untoggle filter on `is_favorite True/False`)

PNL0 replied:
Done. 
For this I made a filter very similar to the pause filter (with the options all, favorite and unfavorite).

PNL0 committed:
Create new filters and update the frontend to use the UI version of get_dags.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/common/parameters.py:
```diff
@@ -112,6 +114,35 @@ def depends(cls, offset: NonNegativeInt = 0) -> OffsetFilter:
+                ~select(DagFavorite)
+                .where((DagFavorite.dag_id == DagModel.dag_id) & (DagFavorite.user_id == self.user_id))
+                .exists()
+            )
```
Please don't use bitwise operator, sqlalchemy exposes "not" "and" "or" which are clearer.

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -317,6 +322,64 @@ def patch_dags(
+    session.execute(insert(DagFavorite).values(dag_id=dag_id, user_id=user_id))
+
+    return Response(status_code=status.HTTP_204_NO_CONTENT)
+
```
Please check at other endpoints to see how we achieve `204` no response endpoint.

(no return + declare the response status code + no type in the return annotation, etc.)

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/openapi/v2-rest-api-generated.yaml:
```diff
@@ -3140,6 +3148,104 @@ paths:
+          description: Successful Response
+          content:
+            application/json:
+              schema: {}
```
Spec is wrong that should be `204`  (but if you follow how we do it for 204 no response, this will be corrected)

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Dashboard/FavoriteDags/FavoriteDags.tsx:
```diff
@@ -0,0 +1,72 @@
+      <Flex color="fg.muted" my={2}>
+        <FiStar />
+        <Heading ml={1} size="xs">
+          {translate("favorite.favoriteDags")}
```
Translation need to be changed to `First 10 favorite dags" or something like this. To explicit the fact that not all of the favorite dags will be displayed there but just the first 10 . (You can pass `limit` value as a translation parameter)

PNL0 replied:
Done.

pierrejeambrun reviewed (commented):
Nice, a few suggestions but that's looking great 🎉

PNL0 committed:
Small fixes.

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/Dashboard/FavoriteDags/FavoriteDags.tsx:
```diff
@@ -0,0 +1,73 @@
+
+  useEffect(() => {
+    void queryClient.refetchQueries({ queryKey: ["DagServiceGetDagsUi"] });
+  }, [location.key, queryClient]);
```
What are you trying to do here?

PNL0 replied:
The idea was to guarantee that the dashboard is always updated with the current favorite DAGs, even when you favorite some DAG in another location (DagList for instance).

However, I just tried to remove that code and the feature still works as expected, therefore the code is unnecessary and is removed in my last commit.

bbovenzi commented on airflow-core/src/airflow/ui/src/queries/useFavoriteDag.ts:
```diff
@@ -0,0 +1,48 @@
+      UseDagServiceGetDagKeyFn({ dagId }, [{ dagId }]),
+      UseDagServiceGetDagDetailsKeyFn({ dagId }, [{ dagId }]),
+      UseDagRunServiceGetDagRunsKeyFn({ dagId }, [{ dagId }]),
+      UseTaskInstanceServiceGetTaskInstancesKeyFn({ dagId, dagRunId: "~" }, [{ dagId, dagRunId: "~" }]),
```
I think we actually only need to invalidate the the first key, no? The rest are ok.

PNL0 replied:
Done.

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/DagsList/DagsList.tsx:
```diff
@@ -132,6 +133,12 @@ const createColumns = (
     header: "",
   },
+  {
+    accessorKey: "favorite",
+    cell: ({ row: { original } }) => <FavoriteDagButton dagId={original.dag_id} withText={false} />,
```
```suggestion
    accessorKey: "favorite",
    cell: ({ row: { original } }) => <FavoriteDagButton dagId={original.dag_id} withText={false} />,
    enableHiding: false,
```

PNL0 replied:
Done.

PNL0 committed:
Small fixes.

Co-authored-by: Mariana Isabel Marçal Santana <mariana.marcal.santana@tecnico.ulisboa.pt>

bbovenzi commented on airflow-core/src/airflow/ui/src/components/DagActions/FavoriteDagButton.tsx:
```diff
@@ -0,0 +1,73 @@
+      { dagId },
+      {
+        onSuccess: () => {
+          void refetch();
```
We already invalidate the query in `useFavoriteDag()` I don't think this extra refetch is necessary

PNL0 replied:
Done.

bbovenzi reviewed (commented):
Looks like we need a rebase too

PNL0 committed:
Rebase and small fix

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Dashboard/FavoriteDags/FavoriteDagCard.tsx:
```diff
@@ -0,0 +1,56 @@
+  readonly latestRuns: Array<DAGRunResponse>;
+};
+
+export const FavoriteDagCard = ({ dagId, dagName, latestRuns }: FavoriteDagProps) => (
```
When there is no DagRun yet for the dag, we end up with 
![Screenshot 2025-06-27 at 16 52 06](https://github.com/user-attachments/assets/61b7eabd-dc21-4f98-9fec-c408ab9c1c95)


Maybe put an explicit message that "There is no DagRun for this dag yet". And also make the card heigh full height. It's weird to have a small card like this.

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Dashboard/FavoriteDags/FavoriteDagCard.tsx:
```diff
@@ -0,0 +1,56 @@
+  <Box width="100%">
+    <RouterLink to={`/dags/${dagId}`}>
+      <Button
+        borderRadius="md"
```
![Screenshot 2025-06-27 at 16 54 20](https://github.com/user-attachments/assets/5c240b28-d078-4fac-9b84-9966c8e76d25)

You have an issue here. DagRun tooltip bar already have a `<a>` tag, which is nested inside the global card <a> tag, that can't work like this. Maybe make the card non clickable, dag_name -> goes to dag, and dagrun bars -> link to the dagrun

PNL0 replied:
Done.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Dashboard/FavoriteDags/FavoriteDagCard.tsx:
```diff
@@ -0,0 +1,56 @@
+        gap={2}
+        height="auto"
+        px={4}
+        py={3}
```
On small screens there is an issue.

![Screenshot 2025-06-27 at 16 55 45](https://github.com/user-attachments/assets/5ceb9174-3637-4812-b6f2-a8075045cc4d)

pierrejeambrun replied:
Or when there is 10 favorite.

![Screenshot 2025-06-27 at 16 56 54](https://github.com/user-attachments/assets/11e6e2a3-4efe-4f5a-8399-f1a907cc78d2)

PNL0 replied:
Fixed.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Dag/Header.tsx:
```diff
@@ -115,7 +116,10 @@ export const Header = ({
-          <TogglePause dagDisplayName={dag.dag_display_name} dagId={dag.dag_id} isPaused={dag.is_paused} />
+          <Flex align="center" gap={2}>
+            <TogglePause dagDisplayName={dag.dag_display_name} dagId={dag.dag_id} isPaused={dag.is_paused} />
+            <FavoriteDagButton dagId={dag.dag_id} withText={false} />
```
Here we should probably show text.

This is what we are doing for all other buttons in the header. (so we see a clear "* Favorite" and "* Unfavorite" button. Also it should probably be on the right align with the 'reparse' button.

PNL0 replied:
Done.

pierrejeambrun reviewed (commented):
This is looking really promising, and I think users will love it.

A few suggestions and things to fix before we can merge


We need to fix it for `FabAuthManager`. If you run breeze with `--auth-manager FabAuthManager` you will see that the server returns error 500. (apparently we need to cast the `user_id` at some point because SQLAchemy is not happy about it)

PNL0 committed:
Fix issues

pierrejeambrun reviewed (commented):
CI is failing for compat test because "DagFavorite" does not exist in older airflow versions. (provider compat tests)

In the test when we clean the table for dag favorite, we need to make that conditional. If we are on old airflow version don't try to clear that non existing table, if we are on a version >= 3.1.0 we know the table exists so we can clear it.

pierrejeambrun reviewed (approved):
Super cool! 🎉 

Thanks for the hard work.

We just need to fix the small backcompat providers tests and we should be ready to merge.

bbovenzi reviewed (approved):
lgtm whenever CI is green

Looks like the codegen API files are being formatted when they shouldn't be

PNL0 committed:
Fixed error in CI pipeline

PNL0 committed:
Fixed errors in CI pipeline.

PNL0 committed:
Fix CI pipeline issues.

pierrejeambrun reviewed (commented):
There's a conflict on a generated file (because something must have been merged in main). 

We need one more rebase and regeneration of front-end types and we should be good to merge.
```bash
pre-commit run ts-compile-format-lint-ui --all-files
```

PNL0 committed:
Regeneration of front-end types
